PovioKit/Classes/Networking/RestClient/RestClient.swift

```diff
@@ -0,0 +1,160 @@
+  func request(endpoint: EndpointProtocol,
+               method: HTTPMethod,
+               parameters: RestClientProtocol.Params? = nil,
+               headers: Headers? = nil,
+               _ result: RestClientProtocol.DataResult) {
+    updateToken { [weak self] token in
+      guard let strongSelf = self else { return }
+      Logger.debug("\(method.rawValue) (started): " + endpoint.path, params: ["url": endpoint.url])
+      let requestHeaders = self?.combineRequestHeaders(customHeaders: headers, token: token)
+      let encoding: ParameterEncoding = method == .get ? URLEncoding.default : JSONEncoding.default
+      strongSelf.manager
+        .request(endpoint.url, method: method, parameters: parameters, encoding: encoding, headers: requestHeaders)
+        .validate(statusCode: 200..<300)
+        .responseData { response in
+          let statusCode = response.response?.statusCode ?? 0
+          switch response.result {
+          case .success(let data):
+            Logger.debug("\(method.rawValue) (success): " + endpoint.path, params: ["status": statusCode, "url": endpoint.url])
+            let dataResponse = DataResponse(statusCode: response.response?.statusCode ?? 0, data: data)
+            strongSelf.retryCounter.clear(for: endpoint)
+            result?(.success(dataResponse))
+          case .failure(let error):
+            Logger.debug("\(method.rawValue) (failure): " + endpoint.path, params: ["status": statusCode, "url": endpoint.url])
+            switch strongSelf.retryCounter.shouldRetry(for: endpoint) {
+            case true:
+              let retryCount = strongSelf.retryCounter.retry(for: endpoint)
+              Logger.debug("\(method.rawValue) (retry): " + endpoint.path, params: ["retryAttempt": retryCount, "url": endpoint.url])
+              strongSelf.request(endpoint: endpoint, method: method, parameters: parameters, headers: headers, result)
+            case false:
+              strongSelf.retryCounter.clear(for: endpoint)
+              result?(.failure(.generic(error)))
+            }
+          }
+      }
+    }
+  }
+  
+  func updateToken(_ completion: ((_ token: String?) -> Void)?) {
+    guard let currentUser = AccountManager.shared.user.entity?.firebaseUser else {
+      completion?(nil)
+      return
+    }
+    // this will retrieve token, if expired a refresh one is generated automatically
+    currentUser.getIDToken { (idToken, _) in
+      completion?(idToken)
+    }
+  }
+  
+  func combineRequestHeaders(customHeaders: Headers?, token: String?) -> Headers {
+    var headers: Headers = Alamofire.SessionManager.defaultHTTPHeaders
+    
+    // authorization header
+    if let token = token {
+      headers["Authorization"] = "Bearer \(token)"
+    } else {
+      headers["Authorization"] = nil
+    }
+    
+    // app specific headers
+    headers["X-SKIP-APP-PLATFORM"] = "iOS"
+    headers["X-SKIP-APP-NAME"] = "renter"
+    headers["X-SKIP-APP-VERSION-STRING"] = UIDevice.appVersion
```

Comment:
Remove app specific headers. We should also define generic headers at the setup stage.